src/stactools/core/utils/raster_footprint.py

```diff
@@ -18,6 +21,622 @@
+        self.simplify_tolerance = simplify_tolerance
+        self.no_data = no_data
+
+    def footprint(self) -> Optional[Dict[str, Any]]:
+        """Produces the footprint surrounding data (not nodata) pixels in
+        the source image. If the footprint was unable to be computed, None
+        is returned.
+
+        Returns:
+            Optional[Dict[str, Any]]: A GeoJSON dictionary containing the
+            footprint polygon.
+        """
+        mask = self.data_mask()
+        polygon = self.data_extent(mask)
+        if polygon is None:
+            return None
+        polygon = self.densify_polygon(polygon)
+        polygon = self.reproject_polygon(polygon)
+        polygon = self.simplify_polygon(polygon)
+        return mapping(polygon)  # type: ignore
+
+    def data_mask(self) -> npt.NDArray[np.uint8]:
+        """Produces a mask of valid data in the source image. Nodata pixels
+        values are set to 0, data pixels are set to 1.
+
+        Returns:
+            numpy.NDArray[numpy.uint8]: An 2D array containing 0s and 1s for
+            nodata/data pixels.
+
+        """
+        shape = self.data_array.shape
+        if self.no_data is not None:
+            mask: npt.NDArray[np.uint8] = np.full(shape, 0, dtype=np.uint8)
+            if np.isnan(self.no_data):
+                mask[~np.isnan(self.data_array)] = 1
+            else:
+                mask[np.where(self.data_array != self.no_data)] = 1
+            mask = np.sum(mask, axis=0, dtype=np.uint8)
+            mask[mask > 0] = 1
+        else:
+            mask = np.full(shape, 1, dtype=np.uint8)
+        return mask
+
+    def data_extent(self, mask: npt.NDArray[np.uint8]) -> Optional[Polygon]:
+        """Produces the data footprint in the native CRS.
+
+        Args:
+            mask (numpy.NDArray[numpy.uint8]): A 2D array containing 0s and 1s for
+                nodata/data pixels.
+
+        Returns:
+            Optional[Polygon]: A native CRS polygon of the convex hull of data
+            pixels.
+        """
+        data_polygons = [
+            shape(polygon_dict)
+            for polygon_dict, region_value in rasterio.features.shapes(
+                mask, transform=self.transform
+            )
+            if region_value == 1
+        ]
+
+        if not data_polygons:
+            return None
+        elif len(data_polygons) == 1:
+            polygon = data_polygons[0]
+        else:
+            polygon = MultiPolygon(data_polygons).convex_hull
+
+        return polygon
+
+    def densify_polygon(self, polygon: Polygon) -> Polygon:
+        """Adds vertices to the footprint polygon in the native CRS using either
+        ``self.densification_factor`` or ``self.densification_distance``.
+
+        Args:
+            polygon (Polygon): Footprint polygon in the native CRS.
+
+        Returns:
+            Polygon: Densified footprint polygon in the native CRS.
+        """
+        if self.densification_factor is not None:
```

Comment:
```suggestion
        assert not (self.densification_factor and self.densification_distance)
        if self.densification_factor is not None:
```

This is a good place for an assertion. We do a check w/ a meaningful error in the constructor, but we still want to be sure we haven't set both -- `assert` is appropriate b/c it's cheap, and the only way we trigger it is if someone did something unexpected.

Reply:
Done.